### DISCUSSION TITLE
"Access denied" error with FUSE mountpoint created using a sample Dokany Fuse Windows program using MSVC by linking against "dokanfuse2.lib"

### DISCUSSION THREAD

**Discussion by dipujoseph-revenera** [dokan-dev/dokany#1249] [Q&A] [answered]
Hi all,
I built a sample Dokany Fuse Windows program using Microsoft Visual C++ by linking against the "dokanfuse2.lib" file (taken from the "C:\Program Files\Dokan\Dokan Library-2.2.0\lib\dokanfuse2.lib" file which got installed as part of "Dokan Library 2.2.0.1000 Bundle").

Then started that sample program using the following command from Administrator-mode Command Prompt (in order to have "W:" as the FUSE mount-point):

.\DokanFuseTest2.exe w -oallow_root -oallow_other -s

I am able to see the "W:" drive but get an "Access denied" error while trying to go inside it. Please find attached the screen-shot.
![Screenshot_Fuse_drive](https://github.com/user-attachments/assets/af70abc5-1d59-4ec6-a3b7-f322b43af1d8)

Also, please find the source code of the sample Dokany Fuse Windows program in the "TestDokanFuse.cpp" file present inside the attached zip file - 
[DokanFuseTest2.zip](https://github.com/user-attachments/files/17554898/DokanFuseTest2.zip)
.

What could I be doing wrong?

Is linking against the "dokanfuse2.lib" file (using MSVC) supported? 

Or, is it supported only with "cygdokanfuse2.dll" file along with the use of the following Cygwin DLLs?
> cygwin1.dll
> cygstdc++-6.dll
> cyggcc_s-seh-1.dll

Please let me know. Any pointers will be much appreciated.

Thank you.

**Comment by Liryna** [comment]
Hi @dipujoseph-revenera ,

Have you been able to run the sample without changes ?
https://github.com/dokan-dev/dokany/blob/45021e00ec54e38b936620f339c8fb22c3f999d2/samples/fuse_mirror/fusexmp.c

Getting an "Access denied" error while trying to access it means somewhere in your implementation, you are either returning access denied (or similar) or there is an issue with the communication between your implementation and dokan (callbacks not called).
I would suggest adding logs to see what is happening and use [Procmon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) to monitor the event status when accessing the drive.

> Is linking against the "dokanfuse2.lib" file (using MSVC) supported?

yes it is

**Comment by dipujoseph-revenera** [comment]
Thank you very much @Liryna for responding (including confirmation of support for "dokanfuse2.lib" using MSVC) and for providing your valuable inputs.

I will check regarding the points you mentioned and get back to you soon (after the holiday weekend here in India). 
Thank you!

**Reply by Liryna** [reply]
By the way if you are able to mount, it means the usage of the dokan library works but it is the communication after that has an issue

**Reply by dipujoseph-revenera** [reply]
Thanks @Liryna for the pointers.

I seem to be having **problems with the communication between my implementation and Dokan**. 

What could be the possible reasons for communication issues?

**Reply by dipujoseph-revenera** [reply]
In "ProcMon", I see the following event (details pasted below) for the "Explorer.EXE" process while trying to access the drive mounted by the sample "fusexmp" program.

> Date:	11/8/2024 3:51:39.4118759 PM
> Thread:	81712
> Class:	File System
> Operation:	QueryRemoteProtocolInformation
> Result:	INVALID PARAMETER
> Path:	C:\Windows
> Duration:	0.0000023

Does that give any indications?

Thank you.

**Comment by dipujoseph-revenera** [comment]
Hi @Liryna,
I ran the following command using the "fusexmp.exe" (built using the ".cpp" file in the attached zip file - 
[fusexmp.zip](https://github.com/user-attachments/files/17703448/fusexmp.zip)
 ):

> C:\Dipu\fusexmp\fusexmp\fusexmp\x64\Debug>.\fusexmp.exe o -d -oallow_root -oallow_other -s > fusexmp.log 2>&1
> 
> C:\Dipu\fusexmp\fusexmp\fusexmp\x64\Debug>
> 

Please find the generated debug log file ("fusexmp.log") in the attached zip file -  
[fusexmp_log.zip](https://github.com/user-attachments/files/17703474/fusexmp_log.zip)

The above-mentioned log file was saved after trying to access (go inside) the mounted drive using Windows Explorer.

Also, please note that the callback function "xmp_statfs()" did not get invoked anytime during the above-mentioned testing (confirmed using a separate debugging session with appropriate breakpoints set using the Visual Studio debugger).

Please let me know if the attached debug log file gives any info or pointers regarding what might be going wrong.

Thanks a lot, in advance, for your valuable inputs.

**Reply by dipujoseph-revenera** [reply, marked as answer]
The main issue was that the callback functions were getting called with Unix-style paths instead of Windows-style paths in the function arguments. Code was added in the callback functions to convert the Unix-style paths to Windows-style paths before using them in places where Windows-style paths were expected.
After this, the XXX_getattr() function and XXX_open() function are working as expected.

Thanks @Liryna for providing the required pointers.

The list of "Gotchas" from https://www.cs.hmc.edu/~geoff/classes/hmc.cs135.201109/homework/fuse/fuse_doc.html were also very helpful.
